### PR TITLE
Get 'compileas' working when using a 'filter "files: ..."' scope.

### PR DESCRIPTION
**Pull request by ratzlaff**
Extends `compileas` to be used in file filters
```lua
  project "main"
      language "C++"

      files  { "src/file.c" }
      filter { "files:src/file.c" }
          compileas "C++"
```

### CODE REVIEW

File: modules/d/tests/test_gmake.lua [hunk 1]
```diff
@@ -105,6 +105,11 @@ endif
 		test.capture [[
 $(OBJDIR)/blah.o: blah.d
 	@echo $(notdir $<)
+ifeq (posix,$(SHELLTYPE))
+	$(SILENT) mkdir -p $(OBJDIR)
+else
+	$(SILENT) mkdir $(subst /,\\,$(OBJDIR))
+endif
```
Comment (samsinsane) [inline modules/d/tests/test_gmake.lua]: This seems to be an unrelated change? I think this is how it used to be but it was causing issues when running multiple jobs.

**Comment by tvandijck** [comment]
LGTM, I'm deferring to someone else to review/test this however.

**Comment by samsinsane** [comment]
To be honest, everything except for the gmake changes LGTM. I don't understand enough about the gmake generator to know if something will break - I'll mark this as approved, but hopefully @starkos or @TurkeyMan can chime in on the gmake changes. (For what it's worth, the gmake2 changes make sense!)

**Comment by ratzlaff** [comment]
@tdesveauxPKFX rebased as requested!
